epub33/core/index.html

```diff
@@ -6082,9 +6082,112 @@ <h4>URLs in the OCF Abstract Container</h4>
 					</div>
 
 					<p> In the <a>OCF Abstract Container</a>, when a file uses a URL string to reference another file in
-						the container, the string MUST be a <a data-cite="url#path-relative-scheme-less-url-string"
-							>path-relative-scheme-less-URL string</a>, optionally followed by <code>U+0023 (#)</code>
-						and a <a data-cite="url#url-fragment-string">URL-fragment string</a>. </p>
+						the container, the string MUST be a 
+						<a data-cite="url#path-relative-scheme-less-url-string">path-relative-scheme-less-URL string</a>, 
+						optionally followed by <code>U+0023 (#)</code> and a 
+						<a data-cite="url#url-fragment-string">URL-fragment string</a>. 
+					</p>
+
+					<p class="note">	
+						The properties of the <a>container root URL</a> are such that whatever the amount
+						of <a data-cite="url/#double-dot-path-segment">double-dot path segments</a> in a URL string (for
+						example, <code>../../../secret</code>), it is parsed to a <a>content URL</a> (and not "leak"
+						outside the container). This avoids potential run-time security issues. 
+						Furthermore, the additional constraint and <a href="#algo-out-of-container">algorithm</a> below 
+						ensure that such potentially problematic URLs can also be detected when checking the EPUB Document.
+					</p>
+
+					<p>To validate a URL string <var>url</var> found in the <a>OCF Abstract Container</a>, the following steps SHOULD return <var>true</var>:</p>
```

Comment:
I'm not sure I like "to validate a URL".

There's more to validation than that single algorithm. A URL is valid if and only if it parses without validation errors, **and** (in EPUB only) it returns true to for this algorithm.

If we say "to validate a URL", it almost sounds like we only consider the EPUB-specific part.

Also, taken as a whole,  "to validate a URL, the following steps SHOULD return true" sound odd to me. Like the "SHOULD" only applies if your objective is to validate the URL. 🤔

Reply:
Ya, this is why I didn't think it was exactly right when I mentioned it in the issue. The best I could get was that this validates any URL is valid to use in an EPUB, not that the URL is valid to the URL standard.

You're not parsing a URL exactly, or strictly confirming that URLs resolve within the container, so you've got me kind of stumped what would be a more apt descriptor for it.

Reply:
Why does this algorithm need to be used with absolute URLs, out of curiosity? Are you just trying to make it input agnostic?

Otherwise, why not restrict it to checking relative paths? You could pin a stronger definition name on it and make it a MUST that it return true. If a URL fails, whether that constitutes an error or warning should be a separate statement from how to check conformance.

Again, just for example, that earlier paragraph about relative paths could additionally say : "Path-relative-scheme-less-URL strings SHOULD resolve to resources in the OCF Container. To verify, the result of running **resolve the location in the OCF Container** with a path-relative-scheme-less-URL string MUST be true."

And the algorithm becomes: "To **resolve the location in the OCF Container** of give a path-relative-scheme-less-URL string _url_, perform the following steps. They return a boolean value."

Reply:
I actually like your proposal, @mattgarrish, but I will wait for @rdeltour's comment before I foolishly change it again (I should not have done it).

As for:

> Why does this algorithm need to be used with absolute URLs, out of curiosity? Are you just trying to make it input agnostic?

The problem is that step 9 may lead to true even if _url_ is relative, namely if the value of `base` is in fact absolute. This is a combination we must account for but this then results in the algorithm being agnostic to the relative or absolute nature of _url_.

Reply:
> Why does this algorithm need to be used with absolute URLs, out of curiosity? Are you just trying to make it input agnostic?

what @iherman said.

The algorithm does **need** to work with absolute URLs **as input**; but after parsing it does need to handle remote (i.e. absolute) URLs as well as container URLs. So it has to work with absolute-URL strings, as `http://example.com/my-remote-url` isn't different than `./my-remote-url` with a `http://example.com` base.

> Otherwise, why not restrict it to checking relative paths?
> You could pin a stronger definition name on it and make it a MUST that it return true.

Restricting the algorithm to relative paths or not is strictly equivalent.

The algorithm is a blanket pass for all absolute URLs. I don't understand why restricting it to relative URL strings makes it stronger.

> If a URL fails, whether that constitutes an error or warning should be a separate statement from how to check conformance.

Not convinced, but I may not understand.

> Again, just for example, that earlier paragraph about relative paths could additionally say : "Path-relative-scheme-less-URL strings SHOULD resolve to resources in the OCF Container. To verify, the result of running resolve the location in the OCF Container with a path-relative-scheme-less-URL string MUST be true."

I'm not convinced this is an improvement over the current status of the PR.

For one thing, "SHOULD resolve to resources in the OCF Container" does not work well with the `base` element (which I know is discouraged, but still).  
Also, the algorithm is only syntactic, it does not check if the URL resolves to a resource.

> And the algorithm becomes: "To resolve the location in the OCF Container of give a path-relative-scheme-less-URL string url, perform the following steps. They return a boolean value."

The above sounds odd to me. I would expect that "resolving a location" returns a location, not a boolean. And again, the algorithm does not resolution, it only checks that double-dot segments are balanced with non-double-dot segments in a URL path string.

All that said, I do not disagree with your original point that we can do better at introducing the algorithm. I thought putting the note above it would be enough, but if you skip the note it makes less sense.

Reply:
> The problem is that step 9 may lead to true even if url is relative

Right, I didn't mean that step 9 gets thrown out. I was just trying to figure out what single goal there is that would unite the algorithm. Simplify the input and see if there's a way of calling the algorithm rather than it just existing apart from anything else.

> does not work well with the base element (which I know is discouraged, but still)

Right, I understand. And, yes, I wasn't thinking of that outcome in the name, but I'm not intending the names as things to use at this point. I'm just trying to understand how to make a common theme, if you want, out of what this algorithm is checking.

What bothers me a lot with the current wording is that there is a "should" with no normative reason attached to it. How would epubcheck report this: [Warning] The url is not true? [Warning] The url does not validate ... to?

Looking at it some more, I'd probably recommend dropping the note formatting on the paragraph above. You could likely take that prose and the algorithm and make them a subsection of 6.1.5 for clearer separation. I'm still not sure what a good title would be -- "Checking for Invalid Paths"?

Maybe then the algorithm could have a name like "To **check a path does not leak outside the OCF Container** ..." ? I still wouldn't put a normative requirement into an algorithm introduction, though. Something else should call it with the requirement (ideally the algorithm would state to issue a warning, but I don't think we want to open that can of worms).

Reply:
What if we unite the algorithm with the definition of the class of URL strings valid in EPUB?

> A string is a **valid absolute container URL string** if it is an absolute-URL-with-fragment string.
>
> A string is a **valid relative container URL string** if it is a path-relative-scheme-less-URL string, optionally followed by U+0023 (#) and a URL-fragment string, and if these steps return true:
>  * [insert algorithm here]
>
> In the OCF Abstract Container, any URL string MUST be a valid absolute container URL string or a valid relative container URL string.

Reply:
That's making a lot more sense to me! 😄

Reply:
That makes a lot of sense to me as well, with one remark: why introducing the **valid absolute container URL string** term? This is just an alias, as far as I can see, of [absolute-URL-with-fragment string](https://url.spec.whatwg.org/#absolute-url-with-fragment-string). 

Besides, the name suggests that the URL is such that its "domain" is the same as the content root URL (ie, more normatively, the URL's scheme, host and port are identical to the content root URL's) but I do not think we want to say that, because we use this for any URL, and that should include external URLs, too.

So, shouldn't the algorithm say:

>
> A string is a **valid relative container URL string** if it is a path-relative-scheme-less-URL string, optionally followed by U+0023 (#) and a URL-fragment string, and if these steps return true:
>  * [insert algorithm here]
>
> In the OCF Abstract Container, any URL string MUST be a [absolute-URL-with-fragment string](https://url.spec.whatwg.org/#absolute-url-with-fragment-string) or a valid relative container URL string.

Or do I miss something?

Reply:
@iherman sure, that works too 😊

Just so we're clear: one drawback of this approach (not @iherman’s last tweak, but the combined definition generally) is that we lose the prose that said "when a file uses a URL string to reference another file in the container, the string MUST be".
So the intra-container cross-referencing aspect is less obvious.

Also, a side comment not specific to the current PR: compared to what we had in EPUB 3.2, we removed the requirement that relative intra-container URLs MUST resolve. Should we create a new issue to discuss that?

Reply:
(Just to save some time for the others; I presume, @rdeltour, you refer to the last paragraph in https://www.w3.org/publishing/epub32/epub-ocf.html#sec-container-iri)

Reply:
@iherman yes! (I should have been more explicit, thanks for the pointers 😊)

Reply:
As for the last comment:

> Also, a side comment not specific to the current PR: compared to what we had in EPUB 3.2, we removed the requirement that relative intra-container URLs MUST resolve. Should we create a new issue to discuss that?

You are right and I would consider it as our bug, insofar as in our editing (ie, #1898) we made a mistake; this is a "stuff happens" category.

However, would it be easy expand a little bit on what we have and add this? E.g.

> In the OCF Abstract Container, any URL string MUST be a absolute-URL-with-fragment string or a valid relative container URL string. If it is the latter, the result of parsing it with the container root URL as base MUST identify a resource within the OCF Abstract Container.

Reply:
> You are right and I would consider it as our bug, insofar as in our editing (ie, #1898) we made a mistake; this is a "stuff happens" category.

Yeah, I was aware from the beginning that #1898 removed that. It's two separate problems IMO: #1898 and this one #1939 are more syntactic (can be checked without full knowledge of the OCF content), and the resolution issue is about actual resources (needs knowledge of all the OCF resources).
That point was mentioned here and there since then (by @mattgarrish and me), but always as side comments 😊.
 
> However, would it be easy expand a little bit on what we have and add this? E.g.
> 
> > (…) If it is the latter, the result of parsing it with the container root URL as base MUST identify a resource within the OCF Abstract Container.

Unfortunately this doesn't work well with explicit remote bases.

With the difference in nature mentioned above, I would prefer —if we want to keep that requirement— to make a separate statement.

All that said, I'm not entirely convinced this requirement is important at the OCF level. For the Package Document it probably make sense to require that relative URLs resolve. But for content documents, less so? For instance if I have an `img` with a missing image, it won't break the EPUB or its interoperability. A MUST may be too strong there. A SHOULD is probably enough.
For other OCF resources, a non-resolving URL may not even be an issue at all?

Reply:
> Unfortunately this doesn't work well with explicit remote bases.

Oh sh...t, I always forget about that one...

> All that said, I'm not entirely convinced this requirement is important at the OCF level. For the Package Document it probably make sense to require that relative URLs resolve. But for content documents, less so? For instance if I have an img with a missing image, it won't break the EPUB or its interoperability. A MUST may be too strong there. A SHOULD is probably enough.
For other OCF resources, a non-resolving URL may not even be an issue at all?

O.k. You just convinced me that this is not something we should handle here, and it should be the topic of a separate issue.

I will go ahead and do the changes we agreed upon (ie, https://github.com/w3c/epub-specs/pull/1939#discussion_r761814432) with the hope that we all agree, and we can merge this PR. Would you raise a separate issue and see if we can agree to simply leave the document as is or add that statement with a _SHOULD_ (I agree with you that a _MUST_ is too strong).

Reply:
I have updated the document, see  latest commit. 

Cc @rdeltour @mattgarrish may we merge?